proposals/0011-inline-spirv.md

```diff
@@ -0,0 +1,347 @@
+
+Some instructions expect a pointer as an operand. This is why the
+`vk::ext_reference` attribute was added. I am not aware of any extensions that
+would need this. I suggestion we deprecate it in
+[proposal 0006](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0006-reference-types.md).
+
+Other instructions require that a type be a literal. This is fairly common. The
+attribute `vk::ext_literal` was added to tell the compiler that is needs to add
+a literal value on the instruction instead of the result id of a load. This type
+of attribute would be meaningless in HLSL generally, and I will have to keep the
+attribute.
+
+I believe this part of the existing inline SPIR-V works well, and should be
+kept. If we feel strongly about naming conventions, we could change the names.
+
+### Types
+
+Some extensions introduce new types. Very few extension add new types. It is
+possible to define and use a spir-v type with the existing inline spir-v, but it
+is awkward to use. You can see a sample in the
+[spv.intrinsictypeInteger.hlsl](https://github.com/microsoft/DirectXShaderCompiler/blob/128b6fd16b449df696a5c9f9405982903a4f88c4/tools/clang/test/CodeGenSPIRV/spv.intrinsicTypeInteger.hlsl)
+test. Some the difficulties with this are:
+
+1.  The definition of the type is split in two. There is a definition of a
+    function with has an arbitrary id and the id for the `OpType*` opcode for
+    the type. The second part calls the function with values for all of the
+    operands to the `OpType*` instruction in spir-v.
+1.  The function call must be reachable from the entry point that will use the
+    type even if it will be used to declare a global variable.
+1.  The arbitrary id that is part of the type function is what is used to
+    declare an object of that type. This means the same function cannot be used
+    to declare two different types. Also, if there are multiple header files,
+    the cannot use the same id for two different types. This can become hard to
+    manage.
+
+I propose we deprecate the old mechanism, and replace it with a new type
+`vk::SpirvType<int OpCode, ...>`. The idea is that the template on the type
+contains the opcode and all of the parameters. The difficulty with this is that
+the operands are not just literal integer values. Sometimes they are another
+type. Then the header file could create a partial instantiation with a more
+meaningful name. For example, if you wanted to declare the types from the
+[SPV_INTEL_device_side_avc_motion_estimation](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_device_side_avc_motion_estimation.html)
+you could have
+
+```
+template<typename ImageType>
+typedef VmeImageINTEL vk::SpirvType</* OpTypeVmeImageINTEL */ 5700, Imagetype>
+typedef AvcMcePayloadINTEL vk::SprivType</* OpTypeAvcMcePayloadINTEL */ 5704>
+```
+
+Then the user could simply use the types:
+
+```
+VmeImageINTEL<Texture2D> image;
```

Comment:
Are you planning on allowing those to interact with current annotations?
(If we have some vendor-specific `Texture4D` type by ex, which needs some `vk::image_format`)
Do we say "if the type is using inline spir-v, then then cannot mix with other attributes? (seems the simplest way, but might need to be specified?)

Reply:
If they are using inline spir-v, then have the ability to set all of the parameters like image format in the template of `vk::SpirvType` and they do not need `vk::image_format`. However, if the attribute is a decoration that should apply to the type, then it could still apply. I do not see any good example now, but I don't want to rule it out.